homeassistant/components/tankerkoenig/__init__.py

```diff
@@ -0,0 +1,204 @@
+        radius = conf[CONF_RADIUS]
+
+        additional_stations = conf[CONF_STATIONS]
+
+        _LOGGER.debug("Fetching data for (%s,%s) rad: %s", latitude, longitude, radius)
+
+        try:
+            data = pytankerkoenig.getNearbyStations(
+                self._api_key, latitude, longitude, radius, "all", "dist"
+            )
+        except pytankerkoenig.customException as err:
+            data = {"ok": False, "message": err, "exception": True}
+        _LOGGER.debug("Received data: %s", data)
+        if not data["ok"]:
+            _LOGGER.error(
+                "Setup for sensors was unsuccessful. Error occurred while fetching data from tankerkoenig.de: %s",
+                data["message"],
+            )
+            return
+
+        # Add stations found via location + radius
+        nearby_stations = data["stations"]
+        if not nearby_stations:
+            _LOGGER.warning(
+                "Could not find any station in range. Try with a bigger radius"
+            )
+        else:
+            for station in data["stations"]:
+                self.add_station(station)
+
+        # Add manually specified additional stations
+        for station_id in additional_stations:
+            try:
+                additional_station_data = pytankerkoenig.getStationData(
+                    self._api_key, station_id
+                )
+            except pytankerkoenig.customException as err:
+                additional_station_data = {
+                    "ok": False,
+                    "message": err,
+                    "exception": True,
+                }
+
+            if not additional_station_data["ok"]:
+                _LOGGER.error(
+                    "Error when adding station %s:\n %s",
+                    station_id,
+                    additional_station_data["message"],
+                )
+                # Clear entity dictionary, so that the platform setup fails / no sensors get loaded in hass
+                self._entities = {}
+            else:
+                self.add_station(additional_station_data["station"])
+
+    @property
+    def entity_list(self):
+        """Get the list of all entities the platform is monitoring."""
+
+        # Get flat list from the dictionary of lists
+        return [
+            sensor_entity
+            for sublist in self._entities.values()
+            for sensor_entity in sublist
+        ]
+
+    def update(self, now=None):
+        """Get the latest data from tankerkoenig.de."""
+        _LOGGER.debug("Fetching new data from tankerkoenig.de")
+        entity_list = list(self._entities.keys())
+        data = pytankerkoenig.getPriceList(self._api_key, entity_list)
+
+        if data["ok"]:
+            _LOGGER.debug("Received data: %s", data)
+            for station_id, station_list in self._entities.items():
+                for station in station_list:
+                    station.new_data(data["prices"].get(station_id))
```

Comment:
See https://developers.home-assistant.io/docs/en/integration_fetching_data.html for our recommended way to fetch and update data.

This looks like there's a single api that provides all entity data, so we should use the `DataUpdateCoordinator`.
https://developers.home-assistant.io/docs/en/integration_fetching_data.html#coordinated-single-api-poll-for-data-for-all-entities

Reply:
This PR was done before the coordinated data updater was introduced into home assistant, so I had no chance of using this when I created the commits.

Nevertheless, I rebased my commits and tried implementing it as requested, to the best of my abilities.

By doing that, I found an inconsistency in the documentation you pointed to:
In the example, the ```available``` property is implemented as ```return not self.coordinator.failed_last_update```, but should be ```return self.coordinator.last_update_success``` instead (I got an error otherwise).